File: generator/src/main/java/vipsffm/DiscoverVipsOperations.java
package vipsffm;

import app.photofox.vipsffm.jextract.GParamSpec;
import app.photofox.vipsffm.jextract.GTypeClass;
import app.photofox.vipsffm.jextract.GTypeInstance;
import app.photofox.vipsffm.jextract.VipsRaw;
import app.photofox.vipsffm.jextract.VipsTypeMap2Fn;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static app.photofox.vipsffm.jextract.VipsRaw.C_INT;
import static app.photofox.vipsffm.jextract.VipsRaw.C_LONG;
import static app.photofox.vipsffm.jextract.VipsRaw.C_POINTER;
import static app.photofox.vipsffm.jextract.VipsRaw.VIPS_ARGUMENT_DEPRECATED;
import static app.photofox.vipsffm.jextract.VipsRaw.VIPS_ARGUMENT_INPUT;
import static app.photofox.vipsffm.jextract.VipsRaw.VIPS_ARGUMENT_MODIFY;
import static app.photofox.vipsffm.jextract.VipsRaw.VIPS_ARGUMENT_OUTPUT;
import static app.photofox.vipsffm.jextract.VipsRaw.VIPS_ARGUMENT_REQUIRED;

public class DiscoverVipsOperations {

    private static final Logger logger = LoggerFactory.getLogger(DiscoverVipsOperations.class);
    private static final Path girPath = Path.of("/opt/homebrew/share/gir-1.0/Vips-8.0.gir");

    public static void main(String[] args) throws Exception {
        DiscoveredOperations discoveredOperations;
        try (var arena = Arena.ofConfined()) {
            discoveredOperations = run(arena);
        }
        var operations = discoveredOperations.operations();

        var numberWithGir = operations.stream().filter(it -> it.gir() != null).count();
        logger.info("found {} operations ({} with GIR info):", operations.size(), numberWithGir);
        operations.forEach(it ->
            logger.info("  {}", it.nickname())
        );
        if (numberWithGir != operations.size()) {
            logger.info("missing GIR info:");
            operations.stream().filter(it -> it.gir() == null).forEach(it ->
                logger.info("  {}", it.nickname())
            );
        }
    }

    public static DiscoveredOperations run(Arena arena) throws IOException {
        VipsRaw.vips_init(arena.allocateFrom("vips-ffm"));

        var gir = discoverGIR();
        var girMethods = gir.operations();

        var baseObjectGtype = VipsRaw.g_type_from_name(arena.allocateFrom("VipsObject"));
        var paramEnumGType = discoverParamEnumGType();

        var candidateOperations = new ArrayList<VipsOperation>();
        var callbackPointerHolder = new MemorySegment[1];

        // recursively walk the tree of types starting from VipsObject
        VipsTypeMap2Fn.Function recursiveCallback = (type, a, b) -> {
            var subtypeNickname = VipsRaw.vips_nickname_find(type).getString(0);

            try {
                var operation = VipsRaw.vips_operation_new(arena.allocateFrom(subtypeNickname));
                if (operation.address() == 0L || operation.equals(MemorySegment.NULL)) {
                    logger.info("skipping {}, couldn't make operation", subtypeNickname);
                    VipsRaw.vips_type_map(type, callbackPointerHolder[0], MemorySegment.NULL, MemorySegment.NULL);
                    return MemorySegment.NULL;
                }
                var description = VipsRaw.vips_object_get_description(operation).getString(0);
                var flags = VipsRaw.vips_operation_get_flags(operation);

                var isDeprecated = (flags & VipsRaw.VIPS_OPERATION_DEPRECATED()) == VipsRaw.VIPS_OPERATION_DEPRECATED();

                logger.info("found {}:", subtypeNickname);
                logger.info(" description: {}", description);

                // not bothered about memory leaks during generation
                var namesPointer = arena.allocate(C_POINTER);
                var flagsPointer = arena.allocate(C_POINTER);
                var numArgsPointer = arena.allocate(C_POINTER);
                VipsRaw.vips_object_get_args(operation, namesPointer, flagsPointer, numArgsPointer);
                var namesArray = namesPointer.get(C_POINTER, 0);
                var flagsArray = flagsPointer.get(C_POINTER, 0);
                var numArgs = (long) numArgsPointer.get(C_INT, 0);
                var args = new ArrayList<VipsOperationArgument>();
                for (var index = 0L; index < numArgs; index++) {
                    var namePointer = namesArray.get(C_POINTER, index * C_POINTER.byteSize());
                    var name = namePointer.getString(0);
                    var argFlags = flagsArray.get(C_INT, index * C_INT.byteSize());

                    var pspecPointer = arena.allocate(C_POINTER);
                    var argumentClassPointer = arena.allocate(C_POINTER);
                    var argumentInstancePointer = arena.allocate(C_POINTER);
                    var result = VipsRaw.vips_object_get_argument(operation, arena.allocateFrom(name), pspecPointer, argumentClassPointer, argumentInstancePointer);
                    if (result != 0) {
                        throw new RuntimeException("failed to get object argument");
                    }
                    var pspec = GParamSpec.reinterpret(pspecPointer.get(C_POINTER, 0), arena, segment -> {});
                    var blurb = VipsRaw.g_param_spec_get_blurb(pspec).getString(0);
                    var paramType = GParamSpec.value_type(pspec);
                    var typeName = VipsRaw.g_type_name(paramType).getString(0);

                    args.add(new VipsOperationArgument(
                        name,
                        argFlags,
                        blurb,
                        GValueType.fromString(typeName),
                        (argFlags & VIPS_ARGUMENT_REQUIRED()) == VIPS_ARGUMENT_REQUIRED(),
                        (argFlags & VIPS_ARGUMENT_INPUT()) == VIPS_ARGUMENT_INPUT(),
                        (argFlags & VIPS_ARGUMENT_OUTPUT()) == VIPS_ARGUMENT_OUTPUT(),
                        (argFlags & VIPS_ARGUMENT_DEPRECATED()) == VIPS_ARGUMENT_DEPRECATED(),
                        (argFlags & VIPS_ARGUMENT_MODIFY()) == VIPS_ARGUMENT_MODIFY(),
                        isParamSpecAnEnum(pspec, paramEnumGType)
                    ));
                }

                if (candidateOperations.stream().noneMatch(it -> it.nickname().equals(subtypeNickname))) {
                    var girMethod = girMethods.stream()
                        .filter(it -> Objects.equals(it.cIdentifier(), "vips_" + subtypeNickname))
                        .findFirst()
                        .orElse(null);
                    if (girMethod == null) {
                        logger.warn("failed to find GIR information for method: {}", subtypeNickname);
                    }
                    candidateOperations.add(new VipsOperation(subtypeNickname, description, args, isDeprecated, girMethod));
                }
            } catch (Throwable t) {
                logger.info("skipping " + subtypeNickname + ", couldn't introspect", t);
            }

            VipsRaw.vips_type_map(type, callbackPointerHolder[0], MemorySegment.NULL, MemorySegment.NULL);
            return MemorySegment.NULL;
        };

        callbackPointerHolder[0] = VipsTypeMap2Fn.allocate(recursiveCallback, arena);
        VipsRaw.vips_type_map(baseObjectGtype, callbackPointerHolder[0], MemorySegment.NULL, MemorySegment.NULL);

        return new DiscoveredOperations(
            candidateOperations.stream().sorted((left, right) -> left.nickname().compareTo(right.nickname())).toList(),
            gir.enums().stream().map(VipsEnum::new).toList()
        );
    }

    public static class GIRRepository {

        private List<GIRNamespace> namespaces = List.of();

        @JacksonXmlProperty(localName = "namespace")
        @JacksonXmlElementWrapper(useWrapping = false)
        public void setNamespaces(List<GIRNamespace> value) {
            namespaces = concat(namespaces, value);
        }

        public List<GIRNamespace> getNamespaces() {
            return namespaces;
        }

        public static class GIRNamespace {
            private String name = "";
            private List<GIRClass> classes = List.of();
            private List<GIRMethod> functions = List.of();
            private List<GIREnumeration> enumerations = List.of();

            @JacksonXmlProperty(isAttribute = true)
            public void setName(String value) {
                name = name + value;
            }

            public String getName() {
                return name;
            }

            @JacksonXmlProperty(localName = "class")
            @JacksonXmlElementWrapper(useWrapping = false)
            public void setClasses(List<GIRClass> value) {
                classes = concat(classes, value);
            }

            public List<GIRClass> getClasses() {
                return classes;
            }

            @JacksonXmlProperty(localName = "function")
            @JacksonXmlElementWrapper(useWrapping = false)
            public void setFunctions(List<GIRMethod> value) {
                functions = concat(functions, value);
            }

            public List<GIRMethod> getFunctions() {
                return functions;
            }

            @JacksonXmlProperty(localName = "enumeration")
            @JacksonXmlElementWrapper(useWrapping = false)
            public void setEnumerations(List<GIREnumeration> value) {
                enumerations = concat(enumerations, value);
            }

            public List<GIREnumeration> getEnumerations() {
                return enumerations;
            }
        }

        public static class GIRClass {
            @JacksonXmlProperty(isAttribute = true)
            public String name = "";

            private List<GIRMethod> constructors = List.of();
            private List<GIRMethod> methods = List.of();

            @JacksonXmlProperty(localName = "constructor")
            @JacksonXmlElementWrapper(useWrapping = false)
            public void setConstructors(List<GIRMethod> value) {
                constructors = concat(constructors, value);
            }

            public List<GIRMethod> getConstructors() {
                return constructors;
            }

            @JacksonXmlProperty(localName = "method")
            @JacksonXmlElementWrapper(useWrapping = false)
            public void setMethods(List<GIRMethod> value) {
                methods = concat(methods, value);
            }

            public List<GIRMethod> getMethods() {
                return methods;
            }
        }

        public record GIRMethod(
            @JacksonXmlProperty(isAttribute = true) String name,
            @JacksonXmlProperty(isAttribute = true, namespace = "c", localName = "identifier") String cIdentifier,
            String doc,
            @JacksonXmlElementWrapper(useWrapping = true) List<GIRParameter> parameters,
            GIRParameter returnValue
        ) {
        }

        public record GIRParameter(
            @JacksonXmlProperty(isAttribute = true) String name,
            @JacksonXmlProperty(isAttribute = true, localName = "transfer-ownership") String transferOwnership,
            @JacksonXmlProperty(isAttribute = true) String direction,
            GIRType type,
            Object varargs
        ) {
        }

        public record GIRType(
            @JacksonXmlProperty(isAttribute = true) String name
        ) {
        }

        public static class GIREnumeration {
            @JacksonXmlProperty(isAttribute = true)
            public String name = "";

            @JacksonXmlProperty(isAttribute = true, namespace = "c", localName = "type")
            public String cType = null;

            public String doc = null;

            private List<GIREnumerationMember> members = List.of();

            @JacksonXmlProperty(localName = "member")
            @JacksonXmlElementWrapper(useWrapping = false)
            public void setMembers(List<GIREnumerationMember> value) {
                members = concat(members, value);
            }

            public List<GIREnumerationMember> getMembers() {
                return members;
            }
        }

        public record GIREnumerationMember(
            @JacksonXmlProperty(isAttribute = true) String name,
            @JacksonXmlProperty(isAttribute = true, namespace = "c", localName = "identifier") String cIdentifier,
            String doc
        ) {
        }

        // matches the merging behaviour Jackson relied on when these models were Kotlin
        // properties with accumulating setters (`field += value`)
        private static <T> List<T> concat(List<T> existing, List<T> additions) {
            var merged = new ArrayList<>(existing);
            merged.addAll(additions);
            return merged;
        }
    }

    record DiscoveredGIRInfo(
        List<GIRRepository.GIRMethod> operations,
        List<GIRRepository.GIREnumeration> enums
    ) {
    }

    private static DiscoveredGIRInfo discoverGIR() throws IOException {
        try (var xmlInputStream = Files.newInputStream(girPath)) {
            var xmlMapper = XmlMapper.builder()
                .defaultUseWrapper(false)
                .configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true)
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
                .build();

            var xml = xmlMapper.readValue(xmlInputStream, GIRRepository.class);
            var vipsNamespaceXml = xml.getNamespaces().stream()
                .filter(it -> it.getName().equals("Vips"))
                .findFirst()
                .orElseThrow();
            var classMethods = vipsNamespaceXml.getClasses().stream()
                .flatMap(it -> it.getMethods().stream())
                .toList();
            var functions = vipsNamespaceXml.getFunctions();
            var allMethods = new ArrayList<>(classMethods);
            allMethods.addAll(functions);
            var enums = vipsNamespaceXml.getEnumerations();

            logger.info("found {} methods, {} in GIR", allMethods.size(), enums.size());
            return new DiscoveredGIRInfo(allMethods, enums);
        }
    }

    private static long discoverParamEnumGType() {
        var paramSpecArrayPointer = VipsRaw.g_param_spec_types();
        for (var i = 0L; i <= 16L; i++) {
            var gtype = paramSpecArrayPointer.get(C_LONG, i * C_LONG.byteSize());
            var name = VipsRaw.g_type_name(gtype).getString(0);
            if (name.equals("GParamEnum")) {
                return gtype;
            }
        }
        throw new RuntimeException("failed to find GParamEnum type - is glib loaded?");
    }

    private static boolean isParamSpecAnEnum(MemorySegment paramSpec, long paramSpecEnumGType) {
        var gclass = GTypeInstance.g_class(paramSpec);
        var gtype = GTypeClass.g_type(gclass);
        if (gtype == paramSpecEnumGType) {
            return true;
        }
        return VipsRaw.g_type_check_instance_is_a(paramSpec, paramSpecEnumGType) == 1;
    }
}

File: generator/src/main/kotlin/vipsffm/DiscoverVipsOperations.kt
package vipsffm

import app.photofox.vipsffm.jextract.GParamSpec
import app.photofox.vipsffm.jextract.GTypeClass
import app.photofox.vipsffm.jextract.GTypeInstance
import app.photofox.vipsffm.jextract.VipsRaw
import app.photofox.vipsffm.jextract.VipsRaw.C_INT
import app.photofox.vipsffm.jextract.VipsRaw.C_LONG
import app.photofox.vipsffm.jextract.VipsRaw.C_POINTER
import app.photofox.vipsffm.jextract.VipsRaw.VIPS_ARGUMENT_DEPRECATED
import app.photofox.vipsffm.jextract.VipsRaw.VIPS_ARGUMENT_INPUT
import app.photofox.vipsffm.jextract.VipsRaw.VIPS_ARGUMENT_MODIFY
import app.photofox.vipsffm.jextract.VipsRaw.VIPS_ARGUMENT_OUTPUT
import app.photofox.vipsffm.jextract.VipsRaw.VIPS_ARGUMENT_REQUIRED
import app.photofox.vipsffm.jextract.VipsTypeMap2Fn
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.MapperFeature
import com.fasterxml.jackson.dataformat.xml.XmlMapper
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import org.slf4j.LoggerFactory
import java.lang.System.getenv
import java.lang.foreign.Arena
import java.lang.foreign.MemorySegment
import java.nio.file.Files
import java.nio.file.Path

data class DiscoveredOperations(
    val operations: List<VipsOperation>,
    val enums: List<VipsEnum>
)

data class VipsOperation(
    val nickname: String,
    val description: String,
    val args: List<VipsOperationArgument>,
    val isDeprecated: Boolean,
    val gir: DiscoverVipsOperations.GIRRepository.GIRMethod?
) {
    override fun toString(): String {
        val gir = if (gir != null) {
            "has gir"
        } else {
            "no gir"
        }
        return "VipsOperation(nickname=$nickname, ${args.size} args, $gir)"
    }
}

data class VipsEnum(
    val gir: DiscoverVipsOperations.GIRRepository.GIREnumeration
)

sealed class GValueType {

    data object Boolean: GValueType()
    data object Int: GValueType()
    data object UInt64: GValueType()
    data object Double: GValueType()
    data object CharArray: GValueType()
    data object Enum: GValueType()
    data object Flags: GValueType()
    data object Object: GValueType()
    data object VipsImage: GValueType()
    data object VipsArrayInt: GValueType()
    data object VipsArrayDouble: GValueType()
    data object VipsArrayImage: GValueType()
    data object VipsRefString: GValueType()
    data object VipsBlob: GValueType()
    data object VipsSource: GValueType()
    data object VipsTarget: GValueType()
    data object VipsInterpolate: GValueType()
    data class Unknown(val rawName: String): GValueType()
}

fun String.toGValueType(): GValueType {
    return when (this) {
        "gboolean" -> GValueType.Boolean
        "gint" -> GValueType.Int
        "guint64" -> GValueType.UInt64
        "gdouble" -> GValueType.Double
        "gchararray" -> GValueType.CharArray
        "GEnum" -> GValueType.Enum
        "GFlags" -> GValueType.Flags
        "GObject" -> GValueType.Object
        "VipsImage" -> GValueType.VipsImage
        "VipsArrayInt" -> GValueType.VipsArrayInt
        "VipsArrayDouble" -> GValueType.VipsArrayDouble
        "VipsArrayImage" -> GValueType.VipsArrayImage
        "VipsRefString" -> GValueType.VipsRefString
        "VipsBlob" -> GValueType.VipsBlob
        "VipsSource" -> GValueType.VipsSource
        "VipsTarget" -> GValueType.VipsTarget
        "VipsInterpolate" -> GValueType.VipsInterpolate
        else -> GValueType.Unknown(this)
    }
}

data class VipsOperationArgument(
    val name: String,
    val flags: Int,
    val blurb: String,
    val type: GValueType,
    val isRequired: Boolean,
    val isInput: Boolean,
    val isOutput: Boolean,
    val isDeprecated: Boolean,
    val isModify: Boolean,
    val isEnum: Boolean
)

private val logger = LoggerFactory.getLogger(DiscoverVipsOperations::class.java)

fun main() {
    val discoveredOperations = Arena.ofConfined().use {
        DiscoverVipsOperations.run(it)
    }
    val operations = discoveredOperations.operations

    val numberWithGir = operations.count { it.gir != null }
    logger.info("found ${operations.size} operations ($numberWithGir with GIR info):")
    operations.forEach {
        logger.info("  ${it.nickname}")
    }
    if (numberWithGir != operations.size) {
        logger.info("missing GIR info:")
        operations.filter { it.gir == null }.forEach {
            logger.info("  ${it.nickname}")
        }
    }
}

object DiscoverVipsOperations {

    private val girPath = Path.of("/opt/homebrew/share/gir-1.0/Vips-8.0.gir")

    fun run(arena: Arena): DiscoveredOperations {
        VipsRaw.vips_init(arena.allocateFrom("vips-ffm"))

        val gir = discoverGIR()
        val girMethods = gir.operations
        val girEnums = gir.enums

        val baseObjectGtype = VipsRaw.g_type_from_name(arena.allocateFrom("VipsObject"))
        val paramEnumGType = discoverParamEnumGType()

        val candidateOperations = mutableListOf<VipsOperation>()
        lateinit var callbackPointer: MemorySegment

        // recursively walk the tree of types starting from VipsObject
        val recursiveCallback = VipsTypeMap2Fn.Function { type, _, _ ->
            val subtypeNickname = VipsRaw.vips_nickname_find(type).getString(0)

            runCatching {
                val operation = VipsRaw.vips_operation_new(arena.allocateFrom(subtypeNickname))
                if (operation.address() == 0L || operation == MemorySegment.NULL) {
                    logger.info("skipping $subtypeNickname, couldn't make operation")
                    VipsRaw.vips_type_map(type, callbackPointer, MemorySegment.NULL, MemorySegment.NULL)
                    return@Function MemorySegment.NULL
                }
                val description = VipsRaw.vips_object_get_description(operation).getString(0)
                val flags = VipsRaw.vips_operation_get_flags(operation)

                val isDeprecated = flags and VipsRaw.VIPS_OPERATION_DEPRECATED() == VipsRaw.VIPS_OPERATION_DEPRECATED()

                logger.info("found $subtypeNickname:")
                logger.info(" description: $description")

                // not bothered about memory leaks during generation
                val namesPointer = arena.allocate(C_POINTER)
                val flagsPointer = arena.allocate(C_POINTER)
                val numArgsPointer = arena.allocate(C_POINTER)
                VipsRaw.vips_object_get_args(operation, namesPointer, flagsPointer, numArgsPointer)
                val namesArray = namesPointer.get(C_POINTER, 0)
                val flagsArray = flagsPointer.get(C_POINTER, 0)
                val numArgs = numArgsPointer.get(C_INT, 0).toLong()
                val args = (0..<numArgs).map { index ->
                    val namePointer = namesArray.get(C_POINTER, index * C_POINTER.byteSize())
                    val name = namePointer.getString(0)
                    val flags = flagsArray.get(C_INT, index * C_INT.byteSize())

                    val pspecPointer = arena.allocate(C_POINTER)
                    val argumentClassPointer = arena.allocate(C_POINTER)
                    val argumentInstancePointer = arena.allocate(C_POINTER)
                    val result = VipsRaw.vips_object_get_argument(operation, arena.allocateFrom(name), pspecPointer, argumentClassPointer, argumentInstancePointer)
                    if (result != 0) {
                        throw RuntimeException("failed to get object argument")
                    }
                    val pspec = GParamSpec.reinterpret(pspecPointer.get(C_POINTER, 0), arena, {})
                    val blurb = VipsRaw.g_param_spec_get_blurb(pspec).getString(0)
                    val type = GParamSpec.value_type(pspec)
                    val typeName = VipsRaw.g_type_name(type).getString(0)

                    VipsOperationArgument(
                        name,
                        flags,
                        blurb,
                        typeName.toGValueType(),
                        isRequired = flags and VIPS_ARGUMENT_REQUIRED() == VIPS_ARGUMENT_REQUIRED(),
                        isInput = flags and VIPS_ARGUMENT_INPUT() == VIPS_ARGUMENT_INPUT(),
                        isOutput = flags and VIPS_ARGUMENT_OUTPUT() == VIPS_ARGUMENT_OUTPUT(),
                        isDeprecated = flags and VIPS_ARGUMENT_DEPRECATED() == VIPS_ARGUMENT_DEPRECATED(),
                        isModify = flags and VIPS_ARGUMENT_MODIFY() == VIPS_ARGUMENT_MODIFY(),
                        isEnum = isParamSpecAnEnum(pspec, paramEnumGType)
                    )
                }

                if (!candidateOperations.any { it.nickname == subtypeNickname }) {
                    val girMethod = girMethods.firstOrNull { it.cIdentifier == "vips_$subtypeNickname"}
                    if (girMethod == null) {
                        logger.warn("failed to find GIR information for method: $subtypeNickname")
                    }
                    candidateOperations += VipsOperation(subtypeNickname, description, args, isDeprecated, girMethod)
                }
            }.getOrElse {
                logger.info("skipping $subtypeNickname, couldn't introspect", it)
            }

            VipsRaw.vips_type_map(type, callbackPointer, MemorySegment.NULL, MemorySegment.NULL)
            MemorySegment.NULL
        }

        callbackPointer = VipsTypeMap2Fn.allocate(recursiveCallback, arena)
        VipsRaw.vips_type_map(baseObjectGtype, callbackPointer, MemorySegment.NULL, MemorySegment.NULL)

        return DiscoveredOperations(
            operations = candidateOperations.sortedBy { it.nickname },
            enums = gir.enums.map { VipsEnum(it) }
        )
    }

    data class GIRRepository(
        @JacksonXmlProperty(localName = "namespace")
        @JacksonXmlElementWrapper(useWrapping = false)
        val namespaces: List<GIRNamespace>
    ) {
        class GIRNamespace {
            @JacksonXmlProperty(isAttribute = true)
            var name: String = ""
                set(value) {
                    field += value
                }

            @JacksonXmlProperty(localName = "class")
            @JacksonXmlElementWrapper(useWrapping = false)
            var classes: List<GIRClass> = listOf()
                set(value) {
                    field += value
                }

            @JacksonXmlProperty(localName = "function")
            @JacksonXmlElementWrapper(useWrapping = false)
            var functions: List<GIRMethod> = listOf()
                set(value) {
                    field += value
                }

            @JacksonXmlProperty(localName = "enumeration")
            @JacksonXmlElementWrapper(useWrapping = false)
            var enumerations: List<GIREnumeration> = listOf()
                set(value) {
                    field += value
                }
        }

        class GIRClass {
            @JacksonXmlProperty(isAttribute = true)
            var name: String = ""

            @JacksonXmlProperty(localName = "constructor")
            @JacksonXmlElementWrapper(useWrapping = false)
            var constructors: List<GIRMethod> = listOf()
                set(value) {
                    field += value
                }

            @JacksonXmlProperty(localName = "method")
            @JacksonXmlElementWrapper(useWrapping = false)
            var methods: List<GIRMethod> = listOf()
                set(value) {
                    field += value
                }
        }

        data class GIRMethod(
            @JacksonXmlProperty(isAttribute = true)
            val name: String,
            @JacksonXmlProperty(isAttribute = true, namespace = "c", localName = "identifier")
            val cIdentifier: String?,
            val doc: String?,

            @JacksonXmlElementWrapper(useWrapping = true)
            val parameters: List<GIRParameter>?,

            val returnValue: GIRParameter?
        )

        data class GIRParameter(
            @JacksonXmlProperty(isAttribute = true)
            val name: String,
            @JacksonXmlProperty(isAttribute = true, localName = "transfer-ownership")
            val transferOwnership: String?,
            @JacksonXmlProperty(isAttribute = true)
            val direction: String?,
            val type: GIRType?,
            val varargs: Unit?
        )

        data class GIRType(
            @JacksonXmlProperty(isAttribute = true)
            val name: String
        )

        class GIREnumeration {
            @JacksonXmlProperty(isAttribute = true)
            val name: String = ""

            @JacksonXmlProperty(isAttribute = true, namespace = "c", localName = "type")
            val cType: String? = null
            val doc: String? = null

            @JacksonXmlProperty(localName = "member")
            @JacksonXmlElementWrapper(useWrapping = false)
            var members: List<GIREnumerationMember> = listOf()
                set(value) {
                    field += value
                }
        }

        data class GIREnumerationMember(
            @JacksonXmlProperty(isAttribute = true)
            val name: String,
            @JacksonXmlProperty(isAttribute = true, namespace = "c", localName = "identifier")
            val cIdentifier: String?,
            val doc: String?,
        )
    }
    data class DiscoveredGIRInfo(
        val operations: List<GIRRepository.GIRMethod>,
        val enums: List<GIRRepository.GIREnumeration>
    )
    private fun discoverGIR(): DiscoveredGIRInfo {
        Files.newInputStream(girPath).use { xmlInputStream ->
            val xmlMapper = XmlMapper.builder()
                .defaultUseWrapper(false)
                .configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true)
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
                .build()
                .registerKotlinModule()

            val xml = xmlMapper.readValue(xmlInputStream, GIRRepository::class.java)
            val vipsNamespaceXml = xml.namespaces.first { it.name == "Vips" }
            val classMethods = vipsNamespaceXml.classes.map { it.methods }.flatten()
            val functions = vipsNamespaceXml.functions
            val allMethods = classMethods + functions
            val enums = vipsNamespaceXml.enumerations

            logger.info("found ${allMethods.size} methods, ${enums.size} in GIR")
            return DiscoveredGIRInfo(allMethods, enums)
        }
    }

    private fun discoverParamEnumGType(): Long {
        val paramSpecArrayPointer = VipsRaw.g_param_spec_types()
        for (i in 0L..16L) {
            val gtype = paramSpecArrayPointer.get(C_LONG, i * C_LONG.byteSize())
            val name = VipsRaw.g_type_name(gtype).getString(0)
            if (name == "GParamEnum") {
                return gtype
            }
        }
        throw RuntimeException("failed to find GParamEnum type - is glib loaded?")
    }

    private fun isParamSpecAnEnum(paramSpec: MemorySegment, paramSpecEnumGType: Long): Boolean {
        val gclass = GTypeInstance.g_class(paramSpec)
        val gtype = GTypeClass.g_type(gclass)
        if (gtype == paramSpecEnumGType) {
            return true
        }
        return VipsRaw.g_type_check_instance_is_a(paramSpec, paramSpecEnumGType) == 1
    }
}